source/_posts/2015-03-26-creating-a-dynamic-single-page-app-for-our-genome-team-using-react.markdown

```diff
@@ -0,0 +1,96 @@
+
+### Choosing a suitable framework 
+
+We spent a day researching different front end frameworks to see what would best fit our needs. Our requirements were:
+- A robust view layer that could work on top of our already-solid Rails backend
+- Performant enough for an interaction-heavy single-page app, with hundreds of editable fields autosaving on change
+- We preferred light, streamlined frameworks, and valued freedom over unnecessary structure
+ 
+After deliberating with other members of our team and the vast online community, we were introduced to [React](http://facebook.github.io/react/), Facebook's view layer framework. We chose React because it provides much-needed structure and support for components in a single page app, without too much boilerplate. 
+ 
+Our plan was to eventually replace all of the existing `*.haml.erb` templates and global coffeescript mixins with discrete React components. We used the [react-rails](https://github.com/reactjs/react-rails) gem, which easily integrates React components with Rails views.
+ 
+In line with the React tutorial, we began by breaking up our UI into functional and visual components. For each component we found the relevant haml template, converted it into jsx and React components using dummy data, and eventually updated it to accept the correct props/state from our top-level component which did all of the dynamic fetching and saving. Then we deleted the associated haml and javascript code and breathed a sigh of relief.
+ 
+### Thinking the React way
+ 
+We love React because it forces you to follow certain [ideological conventions](http://www.reactivemanifesto.org/), but it does not force you into a structure that may not exactly align with your goals.
+ 
+In React, there is an ideal to have a single source of truth. Gone are liberally distributed global event listeners that can conflict and cause pages to get bogged down with transition logic. State is held at the topmost level in React and when state changes, React automatically rerenders only the components that are affected. You can define your components and interactions in a declarative style instead of stringing together possible tranisions triggered by events. Before converting this app to React, we had many bugs around form submission and saving genome progress. However, by modeling state instead of UI transitions, we could easily track changes and save progess incrementally in the background without requiring a page refresh from the user.
```

Comment:
You can define your components and interactions in a declarative style instead of stringing together possible ~~tranisions~~ transitions triggered by events.